29 model training pipeline

Sean1572 opened the pull request:
Adds the whoot_model_training folder to handle easy model training for bioacoustics datasets. Based heavily on https://github.com/UCSD-E4E/pyha-analyzer-2.0/tree/main created by Engineers for Exploration at UCSD. 

Opening PR for feedback for now. I'm going to start conducting some experiments for model training, continue cleaning up the documentation

Sean1572 committed:
Adds inital files for project

Stopping halfway through to get a package manager set up, then once that is set up will return to fill out these files

Sean1572 committed:
Write draft of abstract model class

Stopping halfway through again to address package manager

Sean1572 committed:
Adds loss to ModelOutput

Sean1572 committed:
Merge branch 'dev' into 29_model_training_pipeline

Sean1572 committed:
Update timm_model to include loss function

Sean1572 committed:
Start building dataset handling

Starts a system that ensures some standards in model formation to make training easier to work with. Using Apache Arrow datasets from HuggingFace since I find for large datasets, they are faster than pandas + allow for automatically decoding audio data

Sean1572 committed:
Adds pyha-analyzer as a subdependency

Sean1572 committed:
Adds default format for datasets

Sean1572 committed:
Adds data preprocessing and model format

Sean1572 committed:
Adds the preprocessor for spectrograms

Sean1572 committed:
Adds a demo train script

Sean1572 committed:
Lint

Sean1572 committed:
feat: handle numpy conflict

Sean1572 committed:
Moves files to src and fix bugs

Starting to test system, there were some import statement bugs + running the pipeline so I had to move the files around. Hence the large commit

Sean1572 committed:
Got training working!

Fixed bugs blocking model training, going to test it to make sure it finshes

Next TODOS
- Clean up the codebase
- Connect get the cometML hook working
- Better config options
- Maybe pull some of the codebase into pyha-analyzer-2.0, might be better there...
- Get a model on the data!

Sean1572 committed:
Clean up code

Some linting, documentation, and bug fix for model outputs

Sean1572 committed:
Apply better fix for model input and output

* Adds metrics and successful training!

Was not fully happy with the code, so I decided to rewrite my latest changes at home from scratch, and spend friday morning merging the changes together. This fixes model input / outputs to work with Hugging face more cleanly and allows for metrics to work!

* Add training, Hide Training Artifacts

Sean1572 committed:
Make config easier to manage

Sean1572 committed:
Lint, Spell Check, and Documentation

while I wait for the models to train

Sean1572 committed:
Add environment set up for config

Some comet stuff and settings for changing how many GPUs can be used. Right now just one since distributed training is not working. Probably not putting API key here... I don't want to have people push their APIs to github at all

Sean1572 committed:
Add high level overview of repo

Sean1572 committed:
Add demo of the supplement to the comet ml logging

Sean1572 committed:
Add better model checkpointing

Sean1572 committed:
Update gitignore to hide model_checkpoints

Sean1572 committed:
Added pipeline for augmentations

Sean1572 committed:
Fixed missing run_name after adding custom logging system

Sean1572 committed:
Added confusion matrix

Sean1572 committed:
Linted

Sean1572 committed:
Linted

(yes again, this is going to be a process...)

Sean1572 committed:
Added last of the less destructive linting

Next step, time to try and improve how data gets fed into the models

Sean1572 committed:
Fixed issue formatting model input

Sean1572 committed:
Finalized rounds of linting, mvp for training done

Sean1572 committed:
Add task logging

Sean1572 committed:
Linted after task update

Sean1572 committed:
Linted

Sean1572 committed:
Clean code

Sean1572 committed:
Cleaned code

Sean1572 committed:
Linted

Sean1572 committed:
Add binary extractor for buowset (#39)

* Add binary extractor for buowset

* Adds binary extractor

Sean1572 committed:
Lint the binary model and retest

Sean1572 committed:
Cleans up train script demos

Sean1572 committed:
Lint

kgarwoodsdzwa commented on whoot_model_training/README.md:
```diff
@@ -0,0 +1,33 @@
+To set up environment for model training:
+
+1) run steps 1 - 3 of the installation instructions in `whoot/README.md`
+2) For step 4, specifically run `pip install -e .[model_training, cu128/cpu]`
```
just going over the steps and this command didn't work, i had to install the optional dependencies separately:
pip install -e .[model_training, cu128/cpu]
ERROR: Invalid requirement: 'cu128/cpu]': Expected end or semicolon (after name and no valid version specifier)
    cu128/cpu]
         ^
Hint: It looks like a path. File 'cu128/cpu]' does not exist.

kgarwoodsdzwa replied:
also, running for model_training doesn't get you pyha-analyzer: ERROR: Could not find a version that satisfies the requirement pyha-analyzer; extra == "model-training" (from whoot[model-training]) (from versions: none)
. I needed to pip install uv, and then run uv pip install -e .[model-training] to get it to use the tool.uv.source

Sean1572 replied:
> just going over the steps and this command didn't work, i had to install the optional dependencies separately: pip install -e .[model_training, cu128/cpu] ERROR: Invalid requirement: 'cu128/cpu]': Expected end or semicolon (after name and no valid version specifier) cu128/cpu] ^ Hint: It looks like a path. File 'cu128/cpu]' does not exist.

This could be improved on my side, the command would either be `pip install -e .[model_training, cu128]` OR `pip install -e .[model_training, cpu]`. I tried to show that you could swap between the two, but rereading it now, I can see that isn't clear at all. I'll update the docs for clarity





> also, running for model_training doesn't get you pyha-analyzer: ERROR: Could not find a version that satisfies the requirement pyha-analyzer; extra == "model-training" (from whoot[model-training]) (from versions: none) . I needed to pip install uv, and then run uv pip install -e .[model-training] to get it to use the tool.uv.source

Got it, this is an easy fix, I just need to use the other convention for installing github repos as packages

kgarwoodsdzwa commented on whoot_model_training/whoot_model_training/dataset.py:
```diff
@@ -0,0 +1,84 @@
+            dataset = ds[split]
+            for column in DEFAULT_COLUMNS:
+                state = (
+                    f"The column `{column}` is missing from dataset split `{
```
File "/home/katiegarwood/whoot/whoot_model_training/whoot_model_training/dataset.py", line 46
    f"The column `{column}` is missing from dataset split `{
    ^
SyntaxError: unterminated string literal (detected at line 46)

kgarwoodsdzwa replied:
i fixed this in my own setup by replacing with:
`f"The column `{column}` is missing from dataset split `"
"{split}`. Required by system"`

but unsure if maybe i just have a different version of python or something than you otherwise i would have just pushed my fix. im running on 3.10.12

Sean1572 replied:
Gotcha, this was the result of linting for trying to make the line shorter, I don't think the different Python would break it... I can do something else to clear linting that would be less likely to break it

kgarwoodsdzwa commented on whoot_model_training/whoot_model_training/trainer.py:
```diff
@@ -0,0 +1,97 @@
+        if logger is not None:
+            logger.log_task(training_args.task_name)
+
+        super().__init__(
```
<whoot_model_training.logger.CometMLLoggerSupplement object at 0x71b135a311e0> <class 'whoot_model_training.logger.CometMLLoggerSupplement'>
Traceback (most recent call last):
  File "/home/katiegarwood/whoot/whoot_model_training/train.py", line 162, in <module>
    train(_config)
  File "/home/katiegarwood/whoot/whoot_model_training/train.py", line 131, in train
    trainer = WhootTrainer(
  File "/home/katiegarwood/whoot/whoot_model_training/whoot_model_training/trainer.py", line 88, in __init__
    super().__init__(
  File "/home/katiegarwood/.virtualenvs/birdnet/lib/python3.10/site-packages/pyha_analyzer/trainer.py", line 105, in __init__
    compute_metrics=compute_metrics,
UnboundLocalError: local variable 'compute_metrics' referenced before assignment

kgarwoodsdzwa replied:
i was able to get it to run, but it crashed during this step, and it posted to comet as a failed run.

Sean1572 replied:
Gut feeling says the way you had to install it may do it. This was one of the things I forgot to clean up. I'll address the first comment and see if later stuff has the same issues

> I needed to pip install uv, and then run uv pip install -e .[model-training] to get it to use the tool.uv.source

kgarwoodsdzwa commented on whoot_model_training/configs/config.yml:
```diff
@@ -0,0 +1,10 @@
+# Data paths
+metadata_csv: data/burrowing_owl_dataset/metadata.csv
+data_path: data/burrowing_owl_dataset/audio
+hf_cache_path: data/burrowing_owl_dataset/cache/metadata.hf
```
what's this? i dont think my dataset folder inherently has a cache, is this needed? how can i get it?

Sean1572 replied:
I'll add extra documentation; this should be made automatically. Hugging Face has its caching system for the Apache Arrow Database. It's usually hidden, I like having the path to it more obvious at times. Probably one of the comments worth a discussion

Sean1572 replied:
Just confirmed this is automatically generated, I will add extra documentation about config.yml properties in the model_training/readme.md

Sean1572 committed:
Fixed install

Based on review feedback

Sean1572 committed:
Fixed error handling

Tbh I have no idea if this is the best way to do this. Might look it up later but decided to just do a quick fix

Sean1572 committed:
Update documentation for config.yml

Sean1572 committed:
Renamed extra deps from model_training to model-training

pip does not like underscores...

Sean1572 committed:
Fixed install for pyha

sumi1194 commented:
In whoot_model_training/README.md, I think there was a copy/paste error in step 0 in the 'Running' Section. Can we include this link for reference? https://www.comet.com/docs/v2/guides/experiment-management/configure-sdk/

sumi1194 commented:
I had to add an additional environment variable to specify that the experiment should be logged in the shared conservationtechlab workspace. Without this line, it default uploads to my personal workspace. Can we add this variable to the config file? 
<img width="766" height="258" alt="image" src="https://github.com/user-attachments/assets/e70f8bfd-c0ea-45e4-8eb5-f7b5251062a1" />

Sean1572 commented:
> I had to add an additional environment variable to specify that the experiment should be logged in the shared conservationtechlab workspace. Without this line, it default uploads to my personal workspace. Can we add this variable to the config file?

Good catch, I'll add that in

Sean1572 committed:
Add COMET_WORKSPACE config

Sean1572 committed:
Fix missing comet link in readme.md

Sean1572 committed:
Cleaned config

Sean1572 committed:
Renamed comment to offline to online

Sean1572 committed:
Reworked preprocessors to clarify inheritance

Old code had weird naming conventions caused by iterating during the development processes. Current code generalizes the base_preprocessor to allow for other preprocessors and model_inputs

Sean1572 committed:
Improved model saving with PretrainedModel

kgarwoodsdzwa commented on whoot_model_training/whoot_model_training/data_extractor/buowset_extractor.py:
```diff
@@ -0,0 +1,178 @@
+    ds = ds.cast_column("audio", Audio(sampling_rate=params.sr))
+
+    # Create splits of the data
+    test_ds = ds.filter(lambda x: x["fold"] == params.validation_fold)
```
this is a small error but i just noticed it while doing the esc50 extractor- the test_ds and valid_ds are swapped with what they're defined as, should be test_ds = params.test_fold and valid_ds = validation_fold.

Sean1572 committed:
Fix mispelling of multilabel

kgarwoodsdzwa committed:
60 esc extractor (#68)

* just making this into a branch

copies buowset extractor into a panda one and updated the init

* adding current scripts for esc50 instead of pandas

was working on it on my laptop, but need to commit to be able
to train on a gpu or just more powerful cpu because
my process keeps getting killed by lack of ram it seems

* lowered steps and increased epochs

the model was never evaluating because the default buowset
steps were being used (100) when esc50 is much smaller so it only had
38 steps. So it was appearing like it was not crashing and was doing
the training, but would never get evaluated. It was also not
seeming to learn fast enough with only 2 epochs (more classes than
buowset?) so epochs were increased to 10 which allowed the model
to actually start learning

* fix some pylint, still snake case stuff to fix

* fixed pylint and removed git tracking from the train.py

the train.py is always meant to just be copied and modified by
each user, so it didnt need to be tracked for esc50. the
only differences were the data extractor and the steps and epochs.
pylint done on the extractor

Sean1572 committed:
Swap test_fold and vaild_fold in buowset_extractor.py

Sean1572 committed:
Merge branch 'dev' into 29_model_training_pipeline

Sean1572 committed:
Linted and added more documentation

Sean1572 commented:
Works with buowsetv1.1, already have a few runs ready to go there

Been testing with different timms models as well!

kgarwoodsdzwa commented:
i realized my pylint version was in fact actually not the newest version (was 2.16) so when i upgraded to 3.3.8 it stopped throwing flags about snake case for short variable names, so i think we'll just stick with this.
